Upgrades to dark handling, parallel processing, and adding model of correlation between wavelength bins

emcangi opened the pull request:
## Pull Request Checklist
 - [x] Code Follows PEP8 guidelines (explain any exceptions)
 - [x] New routines don't duplicate existing functionality
 - [x] Original authors of modified routines have been added as reviewers
 
This code mainly includes:
- Update function that creates a .csv file that links lights with darks
- Add a new function that searches a light/dark key .csv file and determines which entries may have had a revision update since the file was made, and returns a new dataframe containing all the entries including updated ones. (must be written to .csv in a higher level call that's not included here--if it makes sense to include it here, I will make a new commit). 
- Improved readability and logic behind determining when dark frames are usable or not, and what to do with the related light frames

I am the original writer of all the routines but would especially like feedback from @planetarymike on 0ce26822926357d40eadd18072617b6873e509fa.

Other changes included here are not large or super important, but helpful for me to merge before completing a new mission reprocess.

emcangi committed:
Update download.py to include a target location for PDS deliveries, for those of us processing data products.

(cherry picked from commit ccf8cf0382a6a9ffa53355fb4f62e616a1e75ea0)

emcangi committed:
Simplify relative_path_from_fname.
Fix typos/paths, delete unnecessary print statement, and clarify docstrings in reprocessing.py.
Clarify comments in echelle.py.
Add explanatory comments in search.py.

emcangi committed:
Start a function that updates filenames in the light/dark key csv.

emcangi committed:
Update the HDU comparison for reprocessed data products by adding logic that just checks if numbers in arrays are "pretty dang close" (to allow for float rounding issues), and printing of any HDUs that are new in the new products.

emcangi committed:
Update the code that makes a master light/dark key csv:
- Remove unneeded comments
- Make sure that date/time is read as a string from the CSV, then formatted as a datetime object
- Add a line to assign the csv to the right folder
- Sort the final data frame
- Save a csv including files with no dark

emcangi committed:
Improve dark frame checking logic so it is more readable.
Add criterion to mark a dark frame as bad if the frame median is super large.
Add a new exception that occurs if both dark frames are bad, which is distinct from light frames being bad.

emcangi committed:
Create folder for fit plots to be saved to if it doesn't exist but saving is requested

emcangi committed:
Create github-actions-demo.yml

Test actions file to start using automation (thanks @dmundra)

emcangi committed:
Add time to allow for IDL to fully compile when opening IDL within the writeout function.

emcangi committed:
Merge branch 'lasp:master' into master

emcangi committed:
Update the update_filenames_in_light_dark_key() function with version from notebook that was successfully used.

emcangi committed:
Clarify name and purpose of a function.

emcangi commented on maven_iuvs/echelle.py:
```diff
@@ -657,21 +657,43 @@ def subtract_darks(light_fits, dark_fits):
-    # check darks are valid
-    if np.isnan(first_dark).any() & np.isnan(second_dark).any():
-        raise Exception(f"Missing critical observation data: no valid darks")
-
```
Please note this logic is still present, but improved to catch more possible reasons they may be bad, see code at right, line 694

planetarymike replied:
line numbers don't match but I see code below under the comment `CHECK VALIDITY OF DARKS`

emcangi committed:
Add some clarifying comments about how early-mission dark files with only one frame are handled.

planetarymike commented on .github/workflows/github-actions-demo.yml:
```diff
@@ -0,0 +1,18 @@
+name: GitHub Actions Demo
```
Is this file really supposed to be a part of this PR? Recommend deletion.

emcangi replied:
No, I was just being lazy

emcangi replied:
425c02f3191cc8329a1472153c6b7504d2a7e92b

planetarymike commented on maven_iuvs/search.py:
```diff
@@ -337,6 +339,8 @@ def get_latest_files(files):
     Prefers highest version number, then production files to stage files,
     and finally highest revision number.
 
+    NOTE: This only works if multiple versions of the files are in the list.
```
I think if multiple versions are not present, this should just return the original list, right? So it 'works', there's just nothing to do.

emcangi replied:
Clarified what I meant 81300aef8b8b0da8e914f3aa17b9ab02d1a8b2ab here

planetarymike replied:
Makes sense, good to clarify the docstrings here.

planetarymike commented on maven_iuvs/search.py:
```diff
@@ -321,6 +320,9 @@ def get_file_version(orbit_number, data_directory,
                            segment=segment,
                            channel=channel,
                            data_directory=data_directory)
+        # TODO: The above call returns a list of strings, but the 
```
Actually, this whole function is gross because it doesn't operate on a file at all but goes to disk to look at what's there. I don't think this function is called by anything else in the repo, recommend deletion.

emcangi replied:
Must have been used by Kyle? But now they have a whole other repository. Anyway, I actually think we should leave it for now. This falls under the issue #50 and I think it should have a separate PR (i.e. we can work on it later). that ok with you?

planetarymike replied:
OK with me to leave it for now

planetarymike commented on maven_iuvs/echelle.py:
```diff
@@ -334,11 +334,163 @@ def downselect_data(index, light_dark=None, orbit=None, date=None, segment=None,
+    problems = {"row": [], "oldname": [], "newname": []}
+
+    # Precompile regex
+    versioned_pat = re.compile(r"^(?P<base>mvn.+?)_(r|s)(?P<rev>\d+)(?P<ext>\.fits\.gz)$")
```
you're actually gonna make everyone who maintains this repo have to learn regular expressions, aren't you??? :sob: :cries due to laziness:

emcangi replied:
https://xkcd.com/208/

emcangi replied:
I added some cheat codes c81201442917ba32f66d598d3376813f2fc7d6b6

planetarymike commented on maven_iuvs/echelle.py:
```diff
@@ -334,11 +334,163 @@ def downselect_data(index, light_dark=None, orbit=None, date=None, segment=None,
+            m = versioned_pat.match(f)
+            if m:
+                f_base = m.group("base")
+                f_rev = int(m.group("rev"))
```
how do you handle stage vs. production files?

planetarymike replied:
Does it prefer production to stage? Would it prefer _r01 to _s03, for example?

emcangi replied:
Well, it was initially treating the copy on disk as the the copy to prefer. But I think that could backfire in rare circumstances. So I rebuilt the whole thing to make it more efficient, compare r vs. s and numbers of revision, and not duplicate code. 43aca59752dfa8dcfe3b41ed84d20b0d3feb7087

emcangi replied:
It now compares r and s, preferring r, and then just picks the biggest revision number available. You may want to have a quick look at the update. I only check to see if the disk file is newer in one of those ways - if the file listed in the light/dark key csv is newer than whatever the user has on disk, nothing is done. (but I don't know how that would be different anyway. It would be weird to ask that routine to go download new files!)

planetarymike replied:
I still feel like this should be able to rely on `get_latest_files` somehow, but it's OK to leave this functionality as separate for now.

planetarymike commented on maven_iuvs/echelle.py:
```diff
@@ -334,11 +334,163 @@ def downselect_data(index, light_dark=None, orbit=None, date=None, segment=None,
+                changes["newname"].append(newest_file)
+
+        # DARKS ------------------------------
+        if dfolder not in dark_folder_cache:
```
dark handling code copy pasted from light handling code above is begging for encapsulation, but if it works it works.

emcangi replied:
This function has been fully rebuilt to address this in 43aca59752dfa8dcfe3b41ed84d20b0d3feb7087

planetarymike commented on maven_iuvs/echelle.py:
```diff
@@ -545,13 +708,15 @@ def coadd_lights(data, n_good):
     n_good : int
              number of valid frames of data used, calculated by subtract_darks.
-
+    exclude_frames : list
```
does not seem to be a parameter in definition above?

planetarymike replied:
also missing in function below... is the docstring change wrong?

emcangi replied:
Hmmm yeah I think this is an artifact. I think it's not needed because I do nansum to ignore any bad values, and n_good should be calculated correctly. I've been testing various quicklooks and they all looked okay. I'll double check tomorrow but for now I've removed this in 2deddcd9215b5960a570053b65ca16ec583551a1

planetarymike commented on maven_iuvs/echelle.py:
```diff
@@ -588,26 +753,50 @@ def subtract_darks(light_fits, dark_fits):
+    # Check that second dark exists - early in mission, we didn't always take 2 darks.
+    # In those cases we should use the first dark.
+    second_dark_exists = True 
+    if np.isnan(second_dark).all():
```
add a comment: `# an all nan second_dark is guaranteed by get_dark_frames() above if only one dark frame exists`

emcangi replied:
added in 84cb3f619af0f201db32140ec823d57714f2519c

planetarymike commented on maven_iuvs/echelle.py:
```diff
@@ -636,22 +825,23 @@ def subtract_darks(light_fits, dark_fits):
-    else:  # First dark is good 
-        if not second_dark_exists: 
+
+        # A bad 1st dark can't be used for light frames 1-n_int in early files
```
I'm not sure we determined that the first dark issue was so bad that we should throw out all the light frames for which no appropriate dark exists. My read of the matching code in the IUVS IDL pipeline suggests that the first dark is used for all frames when no second frame exists.

Dark determination is in the following nested routines:
- `code/superdriver/driver_code/maven_iuvs_process_l1a_to_l1b.pro`
- `code/level1b/iuvs_to_fits_1b.pro`
- `code/level1b/stack_1b_calibrate.pro`
- `code/level1b/iuvs_estimate_dark.pro`

Probably not worth changing to use the IDL procedure now, but maybe worth opening an issue. How many integrations are loss with this treatment?

emcangi replied:
I think very few files have that problem. I agree with you though, I think we should just use the 1st dark. I didn't code that up yet out of caution. It can be an issue.

emcangi replied:
(all ordinals mentioned in this comment assume zero-indexed)

Wait a second - what the comment line you put this comment on is saying is: if there is no 1st dark frame, AND the 0th dark frame is bad/unusable, then there are no valid dark frames. 

Further down in the code, the 0th dark frame---if it's valid and usable---IS used for the 1st through nth light frames if there was only 1 dark frame taken total. 

I think the only remaining issue is in observations with 2 dark frames. In that scenario, if dark 0 is good and dark 1 is bad, lights 1 through n are thrown out. In that case, should we go ahead and use the 0th dark?

planetarymike replied:
where is the 0th dark used? If all of the integers and comments are zero-based, should we change the comments and variable names to reflect this to avoid confusion, or would that only introduce more confusion?

planetarymike replied:
OK, I see it on line 879 below. All good.

planetarymike commented on maven_iuvs/echelle.py:
```diff
@@ -2005,7 +2196,7 @@ def writeout_l1c(light_l1a_path, dark_l1a_path, l1c_savepath, light_fits, fit_pa
                                 text=True, bufsize=1)
         proc.stdin.write(".com write_l1c_file_from_python.pro\n")
-        time.sleep(3) # Be sure it's compiled 
+        time.sleep(5) # Be sure it's compiled 
```
should probably read the output from the terminal to ensure the process has ended, but this can be a todo

emcangi replied:
It has to be read from stdout or stderr or something and it was very hard to get this all working at all, so definitely a todo item

planetarymike reviewed (commented):
Just a few small comments, hopefully super easy to address.

emcangi committed:
Revert "Create github-actions-demo.yml"

This reverts commit f3f0a012cf5c1d9fb8450d6ee89a86136e457614.

emcangi committed:
Clarify comment.

emcangi committed:
Clarify several docstrings, add missing keyword arguments to the docstrings, and rename a variable/align code so that it doesn't overflow the line length limit.

emcangi committed:
Help humans read regular expressions

emcangi committed:
Rebuild the entire function that updates filenames in the light/dark key. The old one was slow and hard to read. That's what I get for vibecoding.

emcangi committed:
Add a suggested comment

emcangi committed:
Remove old arg not needed in docstring

emcangi committed:
Fix inconsistencies in keyword arguments between call to make_light_and_dark_pair_csv and its definition

emcangi committed:
Fix indent mistake

emcangi committed:
Correct bug where wrong module was specified

emcangi committed:
Use 0th dark if 1st dark is present but bad.

emcangi committed:
Add model fit arrays to what will be written out to IDL.

planetarymike commented:
lmk when you're done with this PR--- it looks like you're still committing some changes.

emcangi committed:
Add the functionality to monitor IDL output to determine when scripts are compiled and when running them has successfully completed.

emcangi commented:
@planetarymike Yes, realized I hadn't added the fitted values to the data products, so had to do that. I also decided to add the ability to watch the IDL output to check for completion now... anticipating that if I didn't, I would get tons more hard-to-analyze errors across the mission and have to go back and fix them all, creating more work. 

FYI, the performance has really tanked--up to 3 minutes to generate a single file sometimes. It may be due to the changes made to allow for checking that the script is compiled, etc. There is also the issue of now that we are saving all the model fits to a new HDU, I have it writing out like 5 or 6 .csv files from python, then loading them in IDL. If I had to guess it's probably the first thing slowing it all down - I'll have another go at it later, but for now I'll just let it run slowly. So far it's taken about an hour to do orbits 100-400. Usually it takes minutes.

emcangi committed:
Raise threshhold for median of a dark frame to a value less likely to cause problems on early mission frames.
Add a timestamp passed in from an external script that will be appended to IDL-generated log filenames. IDL is bad at making timestamps that's why I'm doing it here.

emcangi committed:
Add docstrings.

emcangi committed:
Correct a typo

emcangi committed:
Replace the calculation of what used to be called yMRH--index of the row best representing MRH of the observation--using a simpler approach (tested that it produces the same results for different binning schemes), works regardless of the binning scheme.

emcangi committed:
Update convert_l1a_to_l1c() docstring and arguments (prevent process_timestamp from being required so we can run it in notebook sometimes). Make it required if writeout is being done.

emcangi committed:
Fix problems in transmitting nan values from Python to IDL. Brightness HDU information includes both lists of floats and lists of strings, so we have to 1) set the nan values to an int, -9999, so IDL can read them properly; and 2) extend the lists of production_creation_date and oribt_segment to match the lengths of the other lists in that HDU (just repeating the value). If 2) is not done, the value -9999 gets filled in that column which is otherwise strings and causes IDL errors.

emcangi committed:
Catch cruise files when looking for the parent folder of a file.

emcangi committed:
Get rid of unnecessary code

emcangi committed:
Reorganize echelle.py to make it easier to find things. No new code.

emcangi commented:
Had a look at the performance and it looks like the fitting dominates things, and dynesty is still an order of magnitude slower than scipy:

```
# Fit an observation on orbit 724 with all broken or nan frames - Scipy
2025-08-19 11:04:24 Start (t0)
2025-08-19 11:04:24 Dark subtraction: 0.01 seconds (t1) 
2025-08-19 11:04:25 Artifact removal: 0.98 seconds (t2) 
2025-08-19 11:04:27 Finished flattening and fitting: 2.13 seconds (t3)  # <-----------------------
2025-08-19 11:04:27 Opening IDL
2025-08-19 11:04:27 Finished opening IDL and compiling script: 0.16 sec (tb-ta)
2025-08-19 11:04:28 Finished with IDL writeout: 0.45 sec (td-tc)
2025-08-19 11:04:28 Finished writeout with IDL: 0.64 seconds (t4)
```

```
# Fit an observation on orbit 724 with all broken or nan frames - Dynesty
2025-08-19 11:04:28 Start (t0)
2025-08-19 11:04:28 Dark subtraction: 0.01 seconds (t1) 
2025-08-19 11:04:29 Artifact removal: 1.0 seconds (t2) 
2025-08-19 11:05:05 Finished flattening and fitting: 35.83 seconds (t3) # <-----------------------
2025-08-19 11:05:05 Opening IDL
2025-08-19 11:05:05 Finished opening IDL and compiling script: 0.16 sec (tb-ta)
2025-08-19 11:05:05 Finished with IDL writeout: 0.45 sec (td-tc)
2025-08-19 11:05:05 Finished writeout with IDL: 0.64 seconds (t4)
```

```
# Fit an observation on orbit 21494 - Scipy
2025-08-19 10:55:05 Start (t0)
2025-08-19 10:55:05 Dark subtraction: 0.01 seconds (t1) 
2025-08-19 10:55:06 Artifact removal: 1.01 seconds (t2) 
2025-08-19 10:55:08 Finished flattening and fitting: 2.17 seconds (t3)  # <-----------------------
2025-08-19 10:55:08 Opening IDL
2025-08-19 10:55:42 Finished opening IDL and compiling script: 0.17 sec (tb-ta)
2025-08-19 10:55:43 Finished with IDL writeout: 0.47 sec (td-tc)
2025-08-19 10:55:43 Finished writeout with IDL: 0.67 seconds (t4)
```

```
# Fit an observation on orbit 21494 - Dynesty
2025-08-19 10:56:27 Start (t0)
2025-08-19 10:56:27 Dark subtraction: 0.01 seconds (t1) 
2025-08-19 10:56:28 Artifact removal: 0.99 seconds (t2) 
2025-08-19 10:57:04 Finished flattening and fitting: 36.56 seconds (t3)  # <-----------------------
2025-08-19 10:57:04 Opening IDL
2025-08-19 10:57:05 Finished opening IDL and compiling script: 0.16 sec (tb-ta)
2025-08-19 10:57:05 Finished with IDL writeout: 0.48 sec (td-tc)
2025-08-19 10:57:05 Finished writeout with IDL: 0.66 seconds (t4)
```

I'll try 'lazy parallelizing' for now in my bash script that calls Python, since I also still get NaNs in my uncertainties when fitting with Python (tried many combinations of xtol and ftol). With that in mind I think you're good to review this.

emcangi committed:
Clean up some comment lines and print statements

emcangi committed:
Add the ability to time the conversion to l1c.
Modify the transfer CSVs to IDL to be temp files. This required putting several arrays into one CSV in an interleaved fashion.

emcangi committed:
Stop tracking files that don't have a dark separately, and instead just allow the code to return None, and store 'No valid dark found' in the light/dark CSV key.

emcangi committed:
Add light/dark key files to the repo. Current as of August 28, 2025; note that the v14 key only goes through orbit 19964.

emcangi committed:
Add function that will retrieve a matched dark from a key csv containing light and dark pairs.

emcangi committed:
Improve logic in choose_dark so that the dark file closest in time to the given light file, while still occurring earlier than the light (price is right rules), is assigned as the best dark, rather than just picking the first.

emcangi committed:
Use an existing function to find the datetime object from a filenames in get_dark_path rather than regular expressions.
Add a quick bit of error control that prevents the function from breaking if the light passed in is marked as a dark file.

emcangi committed:
Update the v14 light/dark key in two ways:
1) Find rows where lights had multiple dark options, figure out the optimum dark using changes made in 4bc2a9c72904e63b7cb46bb8e078f3ec1630b367, fill in all of the rows for that light with that same dark;
2) Remove all rows that are pure duplicates of one another.

emcangi committed:
Update logic in writeout_l1c() so that the code 1) checks whether the expected files are already written out, OR waits for an update message from IDL that it is finished (which will only be checked if the expected filenames aren't found). This reduces the number of false errors significantly, although it can still produce an error if the expected filename is not right for some reason, or if they haven't been written out by IDL before the check occurs in Python.
Clear the queue when it's done for tidiness.

emcangi committed:
Add trailing slash when returning folder and filename separately in get_dark_path() since no one wants to remember to add slashes.

emcangi committed:
Add missing forward slashes to several lines in the v14 light/dark key, added originally due to careless coding.

emcangi committed:
Add a few more options for filtering data by binning.

emcangi committed:
More explicitly check that both the fits file AND the .xml label file have been written out before allowing skipping to the next file.
Reduce timeout if the IDL message is not detected to avoid things taking forever.

emcangi committed:
add a missing argument description

emcangi committed:
Update light/dark key files with missing periapse observations in orbit block 3100, fix a problem where the wrong folder was recorded in the v14 version, add new files, and generally update filenames according to latest data on disk.

emcangi committed:
Make a bunch of updates relating to light/dark key management:
- Add function that looks for missing files in the past
- Use a folder cache for looking for a match on disk to speed up updating of names
- Make a backup of existing file by default when generating a new CSV
- add csv_path keyword back to the CSV making function for greater control over output file

emcangi committed:
Update get_countrate_diagnostics so that it doesn't fail with nonlinearly binned files

emcangi committed:
Add ability to just use a BU background in a basic fit, rather than having to make a comparison run.
Make determination of dimensions agnostic and based on the datacube passed in to the clean_data routine when "old" cleanup is used (rather than depending on looking up the binning, which isn't needed)

emcangi committed:
Add a missing import statement that was needed in 6affbec264a0b70bf6c1293187059fa3dc80f8ee.

emcangi committed:
Add option to output the dynesty fit plots. This requires the param_bounds to be inside the subroutine that needs them.

emcangi committed:
Fix mistake in log likelihood that was throwing off the chisquared calculations, etc.

emcangi committed:
v15 pipeline script draft before parallelizing further.

emcangi committed:
Fix bugs in light/dark pair file creation:
- Update code so that the new file is overwriting the old one, rather than being saved to the wrong location
- Add some last-ditch code that will search the version of the light/dark key file provided by M. Mayyasi if a dark can't be found - rarely, darks might have been selected manually due to operations idiosyncrasies at the time the data came in. (Note: Assumes a filename, and assumes same folder as the other light/dark keys.)
- Add a line that ensures any duplicate rows created after name updates are removed
- Add version argument to add_new_files_to_light_dark_key so that the newer key won't get updated with the wrong folders
- Update key CSVs accordingly
- Git add old key sent by M. Mayyasi

emcangi committed:
Simplify code that gets lists of .fits.gz and .xml files in a PDS process, add new dates

(cherry picked from commit 454ed9e317f3b355a4f591e76177abe06242366e)

emcangi committed:
Merge remote-tracking branch 'upstream/master'

emcangi committed:
Add delete=False flag and later unlink the temp files once done to ensure they persist long enough for IDL to grab them.

emcangi committed:
Reorganize import statements because they're a mess.

emcangi committed:
Add LSF to the repo.

emcangi committed:
Fix a bug that was producing inf's in the countrate entered into the metadata index.

emcangi committed:
Add option to return high-side boundaries of pixels (this was helpful for diagnosing the bug addressed in e3593fa26b98534151866886a12564a34856e765)

emcangi committed:
Tidy up the log likelihood function.
Remove a wayward negative sign applied to dynesty results that caused the value to be reported incorrectly.
Update the bounds for scipy to ensure they are the same as the dynesty bounds.
Update the scipy tolerances to be something that I determined seems to reproduce the same fits as dynesty, plus or minus a single point of log likelihood.
Fix the value that gets filled in the fit_params dictionary so it has an appropriate label depending on what package was used.
Update the plot_line_fit() routine to account for the new fit_params keys, correctly display the max log likelihood at all times with the correct sign, and compute the correct chisquared and reduced chisquared from the log likelihood.

emcangi committed:
Finalize multiprocessing implementation in RUN_ECHELLE_v15_PIPELINE.py
Update writeout_l1c() in echelle.py to match; specifically, writeout and threading/queue logic now works based on whether cmd_queue is passed in or not.
Update some docstrings, comment strings.
Move scalers for initial guess values above the fitter decision tree since both scipy and dynesty use them.
Add direction to jax to use CPU not GPU.

emcangi committed:
Tidy up comments, commented-out code, remove snark

emcangi committed:
Update run script so that higher level stuff is not run by every worker (caused by 'spawn' logic)

emcangi committed:
Update run script with best dynesty options for accuracy and speed

emcangi committed:
Update the get_wavelengths() function to construct the wavelength array based on the dispersion of the instrument, per previous IDL pipeline. It doesn't really change the fits fortunately.

emcangi committed:
Suppress the printout for dynesty and replace with a final summary

emcangi committed:
Correct bad choices made where different versions of max log likelihood and chi squared were reported depending on the algorithm; make it so everything just reports maxLL and chisquared in the fit_params (after the fitter runs, of course) and then the plot making code doesn't have to do a bunch of weird if statements or compute chisquared within the plot function. Also, correct a spacing issue in loglikelihood() for readability.

emcangi committed:
Add a function to determine if a frame has suspicious data (e.g. if IUVS was observing and the APP began slewing).

emcangi committed:
Delete commented out code missed in 0c3fee2cd9d4e18e13faf0e7cf9055438c2a74bb.

emcangi committed:
Update Dynesty processing code so it's copacetic with multiprocessing usage (when running with multiprocessing, we may open the stderr queue and thread at a higher level, so it may not need to be opened here).
Update the default arguments for livepts and bound to the best values to let Dynesty find a quality solution while running similarly fast as scipy.optimize.minimize.

emcangi committed:
Add checks to weekly report for whether data are present on even and odd orbits

emcangi committed:
Add channel to downselect_data, since it should be useable on folders containing all products, not just echelle (this was relevant when running on AWS).

emcangi committed:
Get rid of timing tracker since I figured out why dynesty was slow and fixed it, and these lines are polluting the files and making them hard to read

emcangi committed:
Add a verbose flag to allow suppressing the dynesty summary message

emcangi committed:
Fix dumb bug

emcangi committed:
Set this variable to a jnp nan JUST in case (I don't think it matters but might as well be consistent)
Also fix a line that's too long

emcangi committed:
Fix code where I used to replace nans with 0s; the change should functionally work out the same, but is more transparent and doesn't violate the principle of surprise or whatever it is

emcangi committed:
Add typical chisquared calculation based on your model values and data values, so we have this one in addition to the one that solves for it starting from the log likelihood value.

emcangi committed:
Oh look, more dark key updates

emcangi committed:
Add ability to distinguish data product version in the quicklooks.
Keep better track of files that just didn't have a dark and call them that so we know it
These are in the same commit because they are all tangled up together

emcangi committed:
whoops this goes with the previous commit that's what I get for commiting at 22:45

emcangi committed:
Send in more stuff as kwargs to more easily control fit plots from higher level functions.
Explicitly close figs in case of processing tons of plots at once
Fix a bug in the kwargs pair_lights_and_darks

emcangi committed:
Improve file comparison routine

emcangi committed:
INDENTS!!!!!!!!1

emcangi committed:
Fix bug where sometimes in noisy files, the algorithm determines the line centers to be outside the wavelength range; in these cases, set the line brightnesses to zero and mark it as a bad frame.

emcangi committed:
Fix bug that was preventing running the code with IDL opened on demand.

emcangi committed:
Add logic to control whether a broken frame is overwritten with NaNs, which is useful for ensuring we never try to fit a bad frame and put it into a data product, but not helpful when doing quicklooks.
Update nan tracking logic in hot pixel cleanup routine accordingly.

emcangi committed:
Add in flag to ensure bad frames are shown in full in quicklooks.
Simplify the functions used to do data cleanup, no reason to import those separately.
Remove warning text since we are now using Dynesty and the uncertainties shouldn't be broken
Ensure the thumbnails are actually the raw data, they were not before despite claims to the contrary

emcangi committed:
Fix code that allows fitting just one integration because the plots returned for it were wrong.

emcangi committed:
Update outlier rejection so the code doesn't nan-out a whole frame that may just have some nans - better preservation of original data.

emcangi committed:
Update some list comprehensions to be more conservative in choice of iterator

emcangi committed:
Fix bug where wrong number of frames were printed on the quicklooks

emcangi committed:
Add some logic to fill in messing Segment fields, because they were missing in the v13 file.

emcangi committed:
Overhaul how the pipeline finds and identifies "bad frames" for both darks and lights:
- Create new functions to find_bad_light_frames(), find_bad_dark_frames(), and get_corrupt_frames() (an especially complicated subset of bad light frames).
- Update get_dark_frames() and subtract_darks() to be simpler and only do what they say they do.
- Remove all logic that builds lists of bad lights in places other than the new function.
- Tidy up nan-ing of fit parameters to account for changes in bad frame processing.
- process the dark files into 2 frames only (get_dark_frames()), and subtract_darks().
- Update variable names to be simpler but sitll clear.
- Update convert_l1a_to_l1c() and quicklook code to account for these changes
- Allow using the 1st (zero-indexed) dark if the 0th dark is not valid.
- Change when exceptions are thrown for two bad darks to allow better overall logic.
- Fix bug in process that checked where fitted central wavelengths were (relates to bad frames, so has to be done here).
- Get rid of unhelpful print statement
- Update compute_ph_per_s_data() to account for new changes in frame management, which affects when a fit is "failed".
- Remove unused imports
- Ensure that darks are not retrieved twice in the quicklook code, because they were previously
- Employ new dictionary explaining light frame fails into the quicklooks which is great and beautiful now
- Remove unneeded arrays in plot_line_fit
- Update docstrings
- Update pix_to_bin to support these changes and fix bug where it was supposed to be able to work on the spectral dimension, but wasn't.

emcangi committed:
Update to nanpercentile. Basic verison was causing problems. Not sure why it didn't happen before but here we are

emcangi committed:
Add functions that let you cache lists of files on your harddrive once so you don't constantly have to go to disk I/O when doing things like full mission reprocess, etc.

emcangi committed:
Move text for MRH to top right corner of thumbnails so it doesn't overlap the messages about frame problems

emcangi committed:
Update code so 'fitter' can be accessed in fit_flat_data and fit_H_and_D.
String cleanup, spacing cleanup.

emcangi committed:
Reorganize echelle.py because it's totally out of hand! Everything in this commit is just moving functions around

emcangi committed:
Fix typos, throw away unused variables

emcangi committed:
Update string because we do handle nonlinearly binned files now

emcangi committed:
Update key

emcangi committed:
Remove accidentally reintroduced print statements.

emcangi committed:
Update the multiprocessing implementation of the l1a->l1c v15 pipeline:
- Add ability to fit a few individual files
- Remove duplicated functions
- Encapsulate some code that launches IDL and compiles so it's modular
- Regularize variable names
- Delete old code/useless strings
- Add new helpful explanatory strings
- Update docstrings

emcangi committed:
Add delete_on_close parameter to the temporary file creation now that Python 3.13 is being used (delete=False should be enough, but this is for safety)

emcangi committed:
Allow 64 bit floats in jax.numpy arrays; without this, jax forces arrays to have dtype=np.float32 (see 'the sharp bits' on jax documentation), which triggers 'RuntimeWarning: overflow encountered in cast" when dynesty.dynamicsampler.py or dynesty.sampler.py tries to assign _LOWL_VAL (which is -1e300) to any nonfinite values of cur_live_logl. Interestingly, this behavior can't be replicated in regular numpy: regular numpy (version 2.4.2) is happy to cast -1e300 to an array with dtype=np.float32, but it does so by silently replacing it with 0.

emcangi committed:
Prevent code from wasting time and generating warnings by checking nan frames for corruption

emcangi committed:
Rename function so it makes more sense what it does.

emcangi committed:
Clarify logic to allow for calling v15 processing of many files in a serialized way, as well as in a parallel way and for one single file.
Modularize function to open IDL and compile script in a better way so that it's more portable into different scripts.
Rename some functions to support the change.
Apply the change to the script to call the  v15 pipeline in parallel.
Correct a typo in a folder.
Add the serial version of the v15 pipeline.

emcangi committed:
Fix bugs in light/dark pair finding:
- In find_match_on_disk(), change re.match to re.search, since re.search looks anywhere in the string but re.match insists on finding the pattern at the start. Use of re.match() was causing some lights to be reported as having no dark, because the code was using the unique ID, which doesn't start at the front of the filename.
- Encapsulate code that returns to the original IDL .sav file and searches it for modularity
- Clarify logic of how darks are looked for - basically, if a filename-looking-string is already in the dark field, we just assume it's good and move on. If anything else is there (like 'None' or 'No valid dark found'), and if the reinvestigate flag is True, then the code will attempt to find a dark again.
- Add a variable to track dark folders, since this needs to be updated if dark names are updated.
- Add some additional regular expressions to help with finding darks in weird cases (some outlimb files, especially around orbits 10300--11700, used outdisk files as their darks. This was obscured in the old pipeline because they had been renamed locally).
- Update the new light/dark keys accordingly

emcangi committed:
Update the regular expression for error messages in processing to allow a wider sequence of words/whitespace between "ERROR:" and the filename.

emcangi committed:
Replace some entries with 'no valid dark' which is correct

emcangi committed:
Update the functions that compare file versions with each other to be much better and more flexible.
Clean up code, delete some unneeded pass statements from old pyhton, fill in _ for unused vars
Ignore empty logs when analyzing results
Ensure error messages matches number of problems reported
Account for files with missing darks correctly

emcangi committed:
Allow looking in other segments for darks
Update dark matching code so that it really chases down the possible darks in all possible places

emcangi committed:
Add some optional logic to download.py to force re-checking of all files to sync because the current logic will not sync a file if the names are the same between VM and local drive but the VM version has a newer modification time. Thus it will miss files that were updated by lower levels of the pipeline if those changes didn't show up in the filename.

emcangi committed:
Fix nested " ' bug

emcangi committed:
Reorganize imports in echelle_graphics so my linter stops yelling at me
Fix spacing

emcangi committed:
Update quicklook code to use the light/dark pair CSVs, make things more readable.

emcangi committed:
Follow PEP8 use enumerate

emcangi committed:
Fix bug where sometimes a bad dark would throw an exception rather than being logged

emcangi committed:
Fix bugs that somehow introduced darks into the light column and used lights in the dark column of the key CSV when grabbing stuff from the old IDL sav file, because some entries in it had two darks for each light, and the first entry was often a light for unknown reasons

emcangi committed:
Add model fit arrays in DN to IDL file writeout.

emcangi committed:
Update quicklook generation code to keep track of any rare scenarios where a light file is missing from the light/dark key file entirely. The quicklook code seems to be more capable of identifying these scenarios because it tries to generate quicklooks for all files indexed in the metadata index, which trolls the folders.

emcangi committed:
Delete obsolete lines

emcangi committed:
Add wavelengths constructed based on the dispersion relation to the CSV writeout for transfer to IDL for l1c file/label write

emcangi committed:
Allow the convert_l1a_to_l1c function to return fit information for convenience

emcangi committed:
Add support for correlation kernel, which describes correlation between adjacent and nearby wavelength bins; photoevents entering a bin are not truly fully independent. Currently only supports 332 wavelength bins (more to be added in next update).

emcangi commented:
Much has been committed but I think this is in a good state for updated review. Unfortunately, fixing issues with darks just led to more and more rabbit holes, and as the pipeline evolved, treatment of darks had to evolve as well in complex ways. 

Here's what's been updated:

1. Dark identification routines and pairing routines have been fixed multiple times.
2. Use of dark frames (e.g. when one or more is bad, when a light frame is bad, etc) has been improved
3. Significant updates to routines that update and keeps updated the light/dark pair files for the whole mission
4. Model fitting results are now written to L1c files, both in physical units (kR/nm) and DN. 
5. The whole pipeline now supports multiprocessing, instigated by RUN_ECHELLE_v15_PIPELINE.py.
6. Transfer of information from Python to csvs that IDL picks up has been finalized.
7. Wavelength grid for fitting is now constructed based on dispersion relation, rather than using what's listed in the l1a files, which per Justin is probably not reliable, and are written out to the l1c files.
8. A function is now available that identifies when the IUVS look direction is well out of the orbital plane.
9. Weekly reports now report data present on even and odd orbits
10. Dynesty processing speed improved to be about equal to that of Scipy processing
11. Explicit functions for chisquared have been added

Overall commits are fairly short so viewing each commit may make the most sense. There is one commit that reorganized code in the middle, so that may make it hard to view everything at once.

planetarymike commented on maven_iuvs/echelle.py:
```diff
@@ -3079,6 +3092,55 @@ def make_fit_param_dict(thelist, is_fitparams=True):
+    # Fill each column of array with the full kernel vector, shifted by i 
+    # places toward the "right" (or end). 
+    for i in range(len(padded_kernel)):
+        kernel_array[:, i] = np.roll(padded_kernel, i)
```
This, in combination with the padded_kernel command on line 3136, will introduce correlation between one side of the spectrum and the other due to the presence of non-zero elements outside the band diagonal. The non-zero elements in the upper right and lower left corners of the matrix need to be set to zero after constructing the correlation matrix with this method.

emcangi replied:
Fixed with https://github.com/lasp/maven_iuvs/pull/83/commits/f6102cec92f445ad56470fd0c5eb7ee18813d7ec

emcangi committed:
Fix bug in correlation kernel with values in the off-diagonal corner extremes

emcangi committed:
Update log likelihood calculation to include log(det(covariance)), but just the plain uncertainties.
Update chisquared calculation to correctly account for changes in loglikelihood
Delete unused chisquared calculation

emcangi committed:
Fix dropped f from fstring
Add tqdm to fit loop to help know what the heck is going on

emcangi committed:
Add new LSF and update all functions to call it by default.
Update some missing docstrings in functions that were updated, since we have to add the argument to the docstrings.

emcangi committed:
Add correlation kernel and associated code to Quicklook code
Update quicklook call to use the new argument structure for fit_H_and_D

emcangi committed:
Generalize variable in graphics.py to be more explicit

planetarymike commented on maven_iuvs/instrument.py:
```diff
@@ -321,9 +330,9 @@ def ran_DN_uncertainty(light_fits, dark_subtracted_and_cleaned_data):
-    ran_DN[np.where(np.isnan(ran_DN))] = 0  # TODO: this is not acceptable lol
+
+    ran_DN = np.sqrt(ran_DN_sq)
+    # Nans can occur in the dark_subtracted_and_cleaned_data term, but that's okay.
```
I’m not sure how a NaN results from the calculations above? Doesn’t seem like any of the functions above can produce a negative value

emcangi replied:
Clarified a lot of docstrings and comments in 1d2c76adc180d889b02514470acc7bb4ca63c7e1 to support this. tl;dr here:

Within this specific function, I believe NaNs can appear if for some reason the ran_DN calculation produces a negative value for ran_DN_sq. The equation is

```
ran_DN_sq = dark_subtracted_and_cleaned_data * fit_function + sigma_background**2
```
fit_function and sigma_background should always be positive values so I believe this can only be negative if some bin within dark_subtracted_and_cleaned_data is negative and with an abs value >  fit_function + sigma_background**2 at the same bin. I don't know how often or if this happens, but it may. 

Outside this function, the pipeline is set up so that any light frames that are 'bad' (not usable in some way, i.e. oversaturated, commanding errors, some bins with NaN DN, no usable associated dark frame) are set to NaN after dark subtraction within the `subtract_darks` function*. This was the easiest way to fit while ignoring garbage frames -- specifically it let me write the code to do all the various operations on the entire data cube without having to do 1) tedious for loops or 2) constantly changing array slicing to ignore the garbage frames.

*Note that it currently doesn't do this in the quicklook code, but I think it probably should do it there also. I haven't reexamined the QL code in a while.

planetarymike commented on maven_iuvs/miscellaneous.py:
```diff
@@ -69,12 +74,22 @@ def mirror_dn_to_deg(dn, inverse=False):
+    value : float
+    price_is_right : bool
+                     If true, finds the nearest entry to v in array that isn't 
+                     larger than value.
```
isn’t this argsort?

emcangi replied:
Unsure about how argsort applies, but I realized there's two extraneous steps here. Reduced the unneeded code and fixed the string which referred to nonexistent variable v. Commit: 58af87d7176e7e3ccae99f7757995b9a01b7bb5f

planetarymike commented on maven_iuvs/search.py:
```diff
@@ -1,18 +1,94 @@
+def build_cache(orbfold=None, cache={}, exclude_xml=True, exclude_dark=True, exclude_txt=True,
+                basefolder="/media/emc/ExtremePro/IUVS/IUVS_Data/l1c_ech_data/FMR_v15/DynestyAWS/"):
+    """
+    Given an orit folder orbfold, will fill the cache with all the files there.
```
typo

emcangi replied:
Haha. 543f552248076afc81e9041c363ebe3b93e0e51e

planetarymike commented on maven_iuvs/statistics.py:
```diff
@@ -1,5 +1,35 @@
+            chi squared value, normal or reduced.
+    """
+
+    chisq = -2 * (LL + (N/2)*np.log(2*math.pi) + 0.5 * logdetcov)
```
does chi-squared extend to multivariate normals will non-zero correlation? Or is it necessarily for independent variates?

emcangi replied:
After thinking about this, I think yes, chi squared with that name is defined to only apply to independent multivariate normal distributions. What I've done here is solve for some quantity that relates the data to the model in the likelihood that we use, and called it chi squared (since it used to be that we were assuming independent gaussian variates). So the name may not be appropriate anymore, but I don't at present know a better way to quantify goodness of fit. 

I think your comment would be addressed by just renaming this 'goodness_of_fit' or similar. Was that your intention?

emcangi replied:
Now covered here e28bcc37fa008cd2acac088be93266ca493b54d1 after we uncovered that chi squared does apply to non independent variates

planetarymike commented on maven_iuvs/echelle.py:
```diff
@@ -892,104 +1201,434 @@ def make_dark_index(ech_l1a_idx):
+        darks[0, :, :] = dark_fits['Primary'].data[0]
+        darks[1, :, :] = dark_fits['Primary'].data[1]
+    elif n_ints_dark > 2:
+        # If there's more than 2, we can just take the element-wise mean of frames 2:end. Ignore nans.
```
Should we avoid NaNs? That is, do NaNs possibly indicate a problem with the dark that means we should throw it out?

emcangi replied:
Probably a good idea to be consistent, fixed here: b2ec21d1dae6191500a1df2c1f09c97aa55d3d0e

planetarymike commented on maven_iuvs/echelle.py:
```diff
@@ -2313,114 +2872,443 @@ def prior_transform(u, param_bounds):
+        if i >= len(kernel):
+            # After a certain point we have to add more zeros
+            mirrored_kernel_padded = np.flip(np.pad(kernel[1:], (0, i-len(kernel[1:]))))
+            kernel_array[:i, i] = mirrored_kernel_padded
```
This method of construction seems much more confusing to understand than the for loop I used in the original notebook, are there really performance considerations that require us to use this more opaque method of construction?

emcangi replied:
No, I think this was a poor fix for the old version. I tested it, your for loop, and the new one in ad4f70a8ee8d32a2e4b6ec288bf312580ac443dc and the new one is the fastest and hopefully simplest to read:

For loop: 2.8 ms ± 34 μs
Old version in 8708a9dcf765a4f0e7ea0533c9bfa49ba1fc640e: 3.92 ms ± 11.8 μs
New version in ad4f70a8ee8d32a2e4b6ec288bf312580ac443dc: 1.47 ms ± 29 μs 

I also made sure the result was the same across all methods. Let me know if this is acceptable! Either way a difference of 1-2 ms on each file is < 1 sec differences across the full mission, so no, no performance requirements here.

planetarymike reviewed (approved):
A few comments, but I didn’t do a detailed review of most of the changes in echelle.py, particularly the dark identification code. If you want me to look at that in detail, let me know!

emcangi committed:
Remove unnecessary steps in price is right rules, fix associated comment string

emcangi committed:
Clarify dark subtraction and bad frame identification.

emcangi committed:
Typo

emcangi committed:
Throw out any frames with nan that appear in situations with > 2 darks (which is rare)

emcangi committed:
Update docstring to be more explicit about assumptions

emcangi committed:
Remove an extraneous internal function (flatten()) and just put the three things it calls directly into the function convert_l1a_to_l1c() for clarity of what's happening. This also matches the behavior in quicklooks code.
Modify ran_DN_uncertainty so it can run either on the raw datacube or a modified datacube (e.g. dark subtracted and cleaned, which is what it was doing before, but this way it's less prescribed and easy to change later if need be)

emcangi committed:
Spacing and comment deletion

emcangi committed:
Add more explanation in log likelihood function to reflect the updated scheme with covariance matrix between bins

emcangi committed:
Add explanatory text that was supposed to be in 1d2c76adc180d889b02514470acc7bb4ca63c7e1

emcangi committed:
Correct bug, can't treat arrays like lists

emcangi committed:
Correct bug in use of MAD because I don't read

emcangi committed:
Update get_kernel_array() to be more legible, shorter, and faster. (this is the fastest of the three methods: the one in commit 8708a9dcf765a4f0e7ea0533c9bfa49ba1fc640e, the one Mike originally wrote, and this one)
